plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitSource.java

```diff
@@ -362,6 +373,11 @@ else if (maxSplitBytes * 2 >= remainingBlockBytes) {
                     splitBytes = internalSplit.getEnd() - internalSplit.getStart();
                 }
 
+                long currentSplitNumber = numberOfProcessedSplits.getAndIncrement();
+                if (currentSplitNumber >= MAX_NUMBER_OF_SPLITS) {
+                    throw new TrinoException(GENERIC_INSUFFICIENT_RESOURCES, format("Number of splits is higher than maximum possible number of splits %d", MAX_NUMBER_OF_SPLITS));
+                }
```

Comment:
Actually, I have second thoughts about that.
 I did not think about this before, but I do not like the fact that current code it limits the number of splits that can be processed by a query, even if we are not doing UPDATE and `HiveUpdatablePageSource` is not being constructed.
What I think would be better, to just record `splitNumber` in `HiveSplit` here (instead `intialRowId` and `maxNumberOfRowsPerSplit`).

Then we can compute do the validation and compute `initialRowId` and `maxNumberOfRowsPerSplit` in `HivePageSourceProvider` where `HiveUpdatablePageSource` is created.
Then we can make the `splitNumber` in `HiveSplit` non-optional. 
And we should move constants definitions to `HivePageSourceProvider`.

WDYT @homar ?

Reply:
sounds really good, thanks